### DISCUSSION TITLE
develop/how_to/process_datasets/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#1712] [Docs]
# develop/how_to/process_datasets/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/develop/how_to/process_datasets/

**Comment by bmorlo** [comment]
Dear all,

thank you for your great work. I happen to have a question.

In my code I am trying to merge two ``sv.DetectionDataset`` objects. Unfortunately, I keep getting the following error:
```
ValueError                                Traceback (most recent call last)

<ipython-input-83-405bb67341d2> in <cell line: 124>()
    122 
    123 # Create the training and the testing detection datasets.
--> 124 detection_dataset_train, IMAGE_READ_INDEX = merged_detection_dataset_from_directory(TRAIN_JSON_DIR_PATH, IMAGE_READ_INDEX)
    125 detection_dataset_test, IMAGE_READ_INDEX = merged_detection_dataset_from_directory(TEST_JSON_DIR_PATH, IMAGE_READ_INDEX)

2 frames

<ipython-input-83-405bb67341d2> in merged_detection_dataset_from_directory(_directory_path, _image_read_index)
    113 
    114     # Merge the datasets
--> 115     merged_detection_dataset = sv.DetectionDataset.merge(datasets_to_merge)
    116 

/usr/local/lib/python3.10/dist-packages/supervision/dataset/core.py in merge(cls, dataset_list)
    311             annotations.update(dataset.annotations)
    312         for dataset in dataset_list:
--> 313             class_index_mapping = build_class_index_mapping(
    314                 source_classes=dataset.classes, target_classes=classes
    315             )

/usr/local/lib/python3.10/dist-packages/supervision/dataset/utils.py in build_class_index_mapping(source_classes, target_classes)
     69     for i, class_name in enumerate(source_classes):
     70         if class_name not in target_classes:
---> 71             raise ValueError(
     72                 f"Class {class_name} not found in target classes. "
     73                 "source_classes must be a subset of target_classes."

ValueError: Class PalletA not found in target classes. source_classes must be a subset of target_classes.
```
Obviously I've tried my best to debug it, but I guess the relation between source and target classes is not clear for me. My feeling is that, because both ``sv.DetectionDataset`` objects have EXACTLY the same class list, i.e. ``['PalletA']``, the merge function runs into an issue.

Could this be a bug inside the merge function? Or is it due to me not encoding class_id and classes correctly when creating the ``sv.DetectionDataset`` objects?

Am I correct, that the "classes" list of a ``sv.DetectionDataset`` object should only comprise a unique list of the found class names?

Thank you so much for your insights!

Best,
Benedikt

**Reply by LinasKo** [reply]
Hi Benedikt,

It's possible that a bug found its way in. What's shown when you print `dataset_1.classes` and `dataset_2.classes`?

**Reply by bmorlo** [reply]
Hi, again!

In both cases, I get ``['PalletA']``. This is correct because both Datasets feature only this one class (a pallet) in their images.

**Reply by LinasKo** [reply]
I can't reproduce it.

Perchance you could share how you're loading the datasets?

Next, would you mind printing the results of this function?

```python
def find_class_mapping(dataset: sv.DetectionDataset):
    class_names = set()
    class_index_name_pairs = set()


    # Existing class names
    for detections in dataset.annotations.values():
        for class_index, class_name in zip(detections.class_id, detections.data.get("class_name", [])):
            class_index_name_pairs.add((class_index, class_name))
    
    # All class names
    for detections in dataset.annotations.values():
        for class_name in detections.data.get("class_name", []):
            class_names.add(class_name)
    return class_index_name_pairs, class_names
```

**Reply by bmorlo** [reply]
Sure,

these are the two functions that I came up with. The first constructs a ``sv.DetectionDataset``. The second function loops through the relevant json files, constructs a ``sv.DetectionDataset`` for each of them, and then merges them at the end. I then call the latter function to create two merged datasets, one for training (will be split later on for validation) and one for testing. 

Apologies if there is an obvious error of mine in the way I construct an ``sv.DetectionDataset``. I will check your function immediately!

```python
def json_file_to_detection_dataset(_json_file_path: str, _image_read_index: int) -> sv.DetectionDataset:

  # Check if the JSON file exists.
  if not os.path.exists(_json_file_path):
      print(f"Error: {_json_file_path} not found")
  try:
      # Open and load the JSON file
      with open(_json_file_path, 'r') as f:
          json_data = json.load(f)
  except json.JSONDecodeError as e:
      # Print an error message if there is a problem decoding the JSON.
      print(f"Error decoding JSON from {_json_file_path}: {e}")

  annotations_as_detections = []
  image_paths_list = []
  classes = []

  for image_item in json_data:
    # Build the current custom annotations dictionary by which more labelling information.
    # from a single image item within the json file can be passed onto the DetectionDataset.
    # Apparently a defaultdict(list) is required. A regular dictionary does not suffice!
    current_custom_annotations = defaultdict(list)

    # Decode the base64 image data.
    image_data = base64.b64decode(image_item['image'].split(',')[1])
    image_path = os.path.join(DATA_DIR_PATH, f"image_{_image_read_index}.jpg")
    image_paths_list.append(image_path)
    with open(image_path, "wb") as f:
      f.write(image_data)

    # Extract bounding box information.
    xmin = image_item["bbox"][0]["xmin"]
    ymin = image_item["bbox"][0]["ymin"]
    xmax = image_item["bbox"][0]["xmax"]
    ymax = image_item["bbox"][0]["ymax"]
    current_class = image_item["bbox"][0]["class"].split("#")[-1]

     # We assume all the JSON files only have "PalletA" class
    current_class_id = 0 #if current_class == "PalletA" else -1 #Assign -1 if class is not PalletA to filter out later if needed.

    current_image_id = image_item["bbox"][0]["id"].split("/")[-1]
    current_height = image_item["height"]
    current_width = image_item["width"]
    # ...
    # There is even more bounding box information to be found in the jsons.

    current_custom_annotations["class_name"].append(current_class)
    current_custom_annotations["image_id"].append(current_image_id)
    current_custom_annotations["height_in_pixels"].append(current_height)
    current_custom_annotations["width_in_pixels"].append(current_width)

    # Depending on the size of the input image, the xyxy values should be scaled.
    current_xyxy = [xmin * int(current_width), ymin * int(current_height), xmax * int(current_width), ymax * int(current_height)]
    classes.append(current_class)

    # For every image, annotations are combined to a sv.Detections object
    annotations_as_detections.append(
        sv.Detections(
            xyxy=np.array([current_xyxy], dtype=np.float32),
            class_id=np.array([current_class_id], dtype=int),
            data=dict(current_custom_annotations)
        )
    )
    # Increase the global image index for reading json images.
    _image_read_index += 1

  # Keep only the unique classes!
  classes = list(set(classes))

  output_detection_dataset = sv.DetectionDataset(classes, image_paths_list, dict(zip(image_paths_list, annotations_as_detections)))
  return output_detection_dataset, _image_read_index

def merged_detection_dataset_from_directory(_directory_path, _image_read_index):
  # Now we generate a DetectionDataset for every json file and merge them.
  json_filenames = []
  for filename in os.listdir(_directory_path):
      if filename.endswith(".json"):  # Check for JSON extension
          json_filenames.append(filename)

  print(f"Found JSON files: {json_filenames}")

  if json_filenames:
    datasets_to_merge = []
    all_classes = []
    for json_file_name in json_filenames:
        current_detection_dataset, _image_read_index = json_file_to_detection_dataset(os.path.join(_directory_path, json_file_name), _image_read_index)
        datasets_to_merge.append(current_detection_dataset)

    # Merge the datasets.
    merged_detection_dataset = sv.DetectionDataset.merge(datasets_to_merge)

  return merged_detection_dataset, _image_read_index

# Create the training and the testing detection datasets.
detection_dataset_train, IMAGE_READ_INDEX = merged_detection_dataset_from_directory(TRAIN_JSON_DIR_PATH, IMAGE_READ_INDEX)
detection_dataset_test, IMAGE_READ_INDEX = merged_detection_dataset_from_directory(TEST_JSON_DIR_PATH, IMAGE_READ_INDEX)
```

**Reply by bmorlo** [reply]
Here is the output from your class mapping function. I called your function right after printing ``dataset.classes`` and before I apply the ``sv.DetectionDataset.merge()`` function.

```
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
Classes in dataset: ['PalletA']
Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
```

**Reply by bmorlo** [reply]
Most interestingly, when I do not supply any additional custom data information when building the ``sv.Detections`` objects like this:

```
annotations_as_detections.append(
        sv.Detections(
            xyxy=np.array([current_xyxy], dtype=np.float32),
            class_id=np.array([current_class_id], dtype=int),
            #data=dict(current_custom_annotations)
        )
    )
```

or, when I simply change ``current_custom_annotations["class_name"].append(current_class)`` to ``current_custom_annotations["name_of_class"].append(current_class)``,

the class mapping results remain empty:

```
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
Classes in dataset: ['PalletA']
Class mapping results: (set(), set())
```

The error, however, persists.

**Reply by LinasKo** [reply]
Thank you, this will be helpful.

**Reply by bmorlo** [reply]
> Most interestingly, when I do not supply any additional custom data information when building the `sv.Detections` objects like this:
> 
> ```
> annotations_as_detections.append(
>         sv.Detections(
>             xyxy=np.array([current_xyxy], dtype=np.float32),
>             class_id=np.array([current_class_id], dtype=int),
>             #data=dict(current_custom_annotations)
>         )
>     )
> ```
> 
> or, when I simply change `current_custom_annotations["class_name"].append(current_class)` to `current_custom_annotations["name_of_class"].append(current_class)`,
> 
> the class mapping results remain empty:
> 
> ```
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> Classes in dataset: ['PalletA']
> Class mapping results: (set(), set())
> ```
> 
> The error, however, persists.

Just so you know, I realized that this behavior is simply due to the way your mapping function is built. I.e. you are checking the custom data supplied by me with the string ``"class_name"``.

**Reply by LinasKo** [reply]
"class_name" is part of `data`, but indeed we give it some special treatment

**Reply by LinasKo** [reply]
@bmorlo, just to clarify - is this still an issue?

**Reply by bmorlo** [reply]
Yes, @LinasKo, unfortunately!

From your point of view, do I generate the ``sv.Detections`` objects correctly within the ``json_file_to_detection_dataset()`` function? Any other suggestion on why I'm having the issue that prevents me from merging two or more ``sv.DetectionDataset`` objects?

Also, I did not fully understand your last reply. I'd appreciate it if you could elaborate on it a bit more. I thought I can supply a dictionary with arbitrary key-value pairs to the ``data`` attribute of the ``sv.Detections`` class.

**Reply by LinasKo** [reply]
Alright, let's figure this out.

What I meant by the last reply is that yes - the datasets need to have a key named "class_name" in the `data`, and the values need to match for merge to succeed.

Just to verify, did you run my debug function before merging took place?
Ideally it should be above the line:

```python
datasets_to_merge.append(current_detection_dataset)
```

**Reply by bmorlo** [reply]
> Here is the output from your class mapping function. I called your function right after printing ``dataset.classes`` and before I apply the ``sv.DetectionDataset.merge()`` function.
> 
> ```
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> Classes in dataset: ['PalletA']
> Class mapping results: ({(0, 'PalletA')}, {'PalletA'})
> ```

Yes, as I showed you here :)

**Reply by bmorlo** [reply]
> Alright, let's figure this out.
> 
> What I meant by the last reply is that yes - the datasets need to have a key named "class_name" in the `data`, and the values need to match for merge to succeed.
> 

Just to make sure I understand you correctly here:

1) I thought the ``data`` paramater was optional in the ``sv.Detections`` constructor.

2) What exactly do you mean by "the values need to match"?

**Reply by LinasKo** [reply]
1. It's optional, in the sense that some detections won't have it. It's not optional for dataset merging purposes. I'm making a note that it's not documented well enough.
2. The sets of classes need to match, and there should be `Detections.empty()` for cases when nothing is detected in an image.

I suspect a bug in how we check against existing classes.

**Reply by bmorlo** [reply]
Okay, then I understand you for 1).

For 2) - how can I, and why should I make sure the sets of classes match? When trying to merge two ``sv.DetectionDatasets`` can I do this only if the dimensions of the datasets match? I.e. should the number of images that are being kept in these datasets with their respective annotations be the same? Or the number of bounding boxes?

When merging, the goal should be to merge datasets with different number of classes and different class names, no?

Maybe I'm missing something obvious: What I try to achieve in my code is to transform ordinary .json files that were generated from some images and include bbox data, to an ``sv.DetectionDataset``. I do this by a) creating an ``sv.Detections`` object with ``xyxy`` and ``class_id`` attributes for each image, then (b) generating a list from all the ``sv.Detections`` objects, and (c) using the encounterer classes, image paths, and the list of ``sv.Detections`` objects to finally construct the ``sv.DetectionDataset``. The datasets that I create like this should then be merged as a last step, to have all the images, their classes, and bounding box data in one place/dataset.

**Reply by LinasKo** [reply]
Ah, apologies again. For 2) I meant the class names and IDs. The number of images shouldn't matter.

> When merging, the goal should be to merge datasets with different number of classes and different class names, no?
In fact, this is not true. The current implementation assumes that very similar datasets are being combined, building towards one that's better for training a specific model.

Merging those with different classes / different numbers of classes also makes sense, and we should add a mechanism that allows it. Currently this is assumed but not documented.

**Reply by bmorlo** [reply]
Thanks! But you've just said that currently, we cannot merge datasets with different numbers of detections, right?

**Reply by LinasKo** [reply]
If that's the case, then I don't understand why the print function only showed `PalletA` as the class. I assumed multiple lines are for each dataset that will be merged.

~~We cannot merge dataset with a different set of class names or class ids.~~ The number of images or annotations doesn't matter - it should work.

~~If you have multiple datasets that contain different classes and try to merge them, it will fail.~~ Let me check the source for a solution.

Edit: I'm wrong. Merging multiple classes should be permitted. I'll need some time.

**Reply by bmorlo** [reply]
Yeah, I was just about to write this: I've seen supervision examples with two datasets containing different numbers of class types/names.

Let me clarify: Every dataset (json file) thar I am workong with and that I transform to a ``sv.DetectionDataset`` currently contains only one type of class - 'PalletA'. In other words each json file comprises multiple images that all show a 'PalletA'. Hence, what the lines that I printed, correspond to the unique lists of class names of every dataset.

Edit: Thank you so much for your support!

**Reply by LinasKo** [reply]
I was wrong initially, but it's still is one of the more unusual bugs I've ever seen. The error should never happen.

Internally, we're collecting all classes from all datasets into a set, and later indexing into it, with classes from one of those datasets. That's when a `ValueError` gets thrown.

<details>
<summary>Mock</summary>

```python
from dataclasses import dataclass

@dataclass
class MockDataset:
    classes: list[str]



def merge_class_lists(class_lists: list[list[str]]) -> list[str]:
    unique_classes = set()

    for class_list in class_lists:
        for class_name in class_list:
            unique_classes.add(class_name)

    return sorted(list(unique_classes))


def build_class_index_mapping(
    source_classes: list[str], target_classes: list[str]
) -> dict[int, int]:
    """Returns the index map of source classes -> target classes."""
    index_mapping = {}

    for i, class_name in enumerate(source_classes):
        if class_name not in target_classes:
            raise ValueError(
                f"Class {class_name} not found in target classes. "
                "source_classes must be a subset of target_classes."
            )
        corresponding_index = target_classes.index(class_name)
        index_mapping[i] = corresponding_index

    return index_mapping


ds_1 = MockDataset(["PalletA"])
ds_2 = MockDataset(["PalletA"])
ds_3 = MockDataset(["PalletA"])
# ds_empty = MockDataset([])
dataset_list = [ds_1, ds_2, ds_3]


classes = merge_class_lists(
    class_lists=[dataset.classes for dataset in dataset_list]
)

for dataset in dataset_list:
    class_index_mapping = build_class_index_mapping(
        source_classes=dataset.classes, target_classes=classes
    )  # Error thrown by this method

    print(class_index_mapping)
```

</details>

Perchance you could run another test? I've expanded the error message to say what target classes it's expecting. If you run your code, it should be more apparent why `PalletA` is not found.

You can install the updated package with:

```bash
pip install git+https://github.com/roboflow/supervision.git@fix/dataset-merge-error
```

Alterantively, you could jump to the place where the error is raised inside the package, and either debug or drop a print statement before the `ValueError` line.

**Reply by bmorlo** [reply]
Alright, I'm using the updated package now. The error persists. I did add your function definitions and your function calls to my code like this:

```ruby
if json_filenames:
    datasets_to_merge = []
    for json_file_name in json_filenames:
        current_detection_dataset, _image_read_index = json_file_to_detection_dataset(os.path.join(_directory_path, json_file_name), _image_read_index)
        datasets_to_merge.append(current_detection_dataset)
    # Added your code.
    classes = merge_class_lists(
    class_lists=[dataset.classes for dataset in datasets_to_merge])

    for dataset in datasets_to_merge:
        # Added your code.
        class_index_mapping = build_class_index_mapping(
        source_classes=dataset.classes, target_classes=classes)  # Error thrown by this method
        print(class_index_mapping)
```
The output is this:
```
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}
{0: 0}

---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

[<ipython-input-46-095ac9a3500f>](https://localhost:8080/#) in <cell line: 137>()
    135 
    136 # Create the training and the testing detection datasets.
--> 137 detection_dataset_train, IMAGE_READ_INDEX = merged_detection_dataset_from_directory(TRAIN_JSON_DIR_PATH, IMAGE_READ_INDEX)
    138 detection_dataset_test, IMAGE_READ_INDEX = merged_detection_dataset_from_directory(TEST_JSON_DIR_PATH, IMAGE_READ_INDEX)

2 frames

[<ipython-input-46-095ac9a3500f>](https://localhost:8080/#) in merged_detection_dataset_from_directory(_directory_path, _image_read_index)
    130 
    131     # Merge the datasets
--> 132     merged_detection_dataset = sv.DetectionDataset.merge(datasets_to_merge)
    133 
    134   return merged_detection_dataset, _image_read_index

[/usr/local/lib/python3.10/dist-packages/supervision/dataset/core.py](https://localhost:8080/#) in merge(cls, dataset_list)
    311         classes = merge_class_lists(
    312             class_lists=[dataset.classes for dataset in dataset_list]
--> 313         )
    314 
    315         annotations = {}

[/usr/local/lib/python3.10/dist-packages/supervision/dataset/utils.py](https://localhost:8080/#) in build_class_index_mapping(source_classes, target_classes)
     69     for i, class_name in enumerate(source_classes):
     70         if class_name not in target_classes:
---> 71             raise ValueError(
     72                 f"Class {class_name} not found in target classes. "
     73                 f"source_classes must be a subset of target_classes ({target_classes})."

ValueError: Class PalletA not found in target classes. source_classes must be a subset of target_classes.
```

**Reply by bmorlo** [reply]
One more interesting remark: When I run the code with your test functions in place, your test functions DO NOT raise the error! It comes from the functions of the normal package - independent of which supervision version (live vs. modified by you) I use.

Also interesting is the fact that the error completely disappears when I supply individual class names (a number increased through a counter...) for every detection/image and the same class_id = 0 for all detections.

When I use below code to alternate between two class types A and B when creating a the detections of a dataset, the result is similar to what I encountered before.

```ruby
    if temp_class_index % 2 == 0:
      current_class = "A"
      temp_class_index += 1
      current_class_id = 0
    else:
      current_class = "B"
      temp_class_index += 1
      current_class_id = 1
```

Here the output:

```
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
{0: 0, 1: 1}
```

**Reply by LinasKo** [reply]
I'm slightly confused, as the error message does not match the error definition.

```
---> 71             raise ValueError(
     72                 f"Class {class_name} not found in target classes. "
     73                 f"source_classes must be a subset of target_classes ({target_classes})."

ValueError: Class PalletA not found in target classes. source_classes must be a subset of target_classes.
```

This should have printed at least the following, in cases where no target classes are found:

```
ValueError: Class PalletA not found in target classes. source_classes must be a subset of target_classes ().
```

**Reply by bmorlo** [reply]
Hmm. I don't understand your point, I'm afraid. I do get this exact error message all the time, no?

**Reply by bmorlo** [reply]
> One more interesting remark: When I run the code with your test functions in place, your test functions DO NOT raise the error! It comes from the functions of the normal package - independent of which supervision version (live vs. modified by you) I use.
> 
> Also interesting is the fact that the error completely disappears when I supply individual class names (a number increased through a counter...) for every detection/image and the same class_id = 0 for all detections.
> 
> When I use below code to alternate between two class types A and B when creating a the detections of a dataset, the result is similar to what I encountered before.
> 
> ```ruby
>     if temp_class_index % 2 == 0:
>       current_class = "A"
>       temp_class_index += 1
>       current_class_id = 0
>     else:
>       current_class = "B"
>       temp_class_index += 1
>       current_class_id = 1
> ```
> 
> Here the output:
> 
> ```
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> {0: 0, 1: 1}
> ```

I believe your opinion on my observations would help me the most. Also, I would like to know whether the output that I'm seeing is in order.

**Reply by LinasKo** [reply]
Sure!

> Hmm. I don't understand your point, I'm afraid. I do get this exact error message, no?

1) Yes. It's perplexing. I added `... ({target_classes})"` to the error specifically so we'd see what `target_classes` are, as that's the root cause of the error. The stack trace of the thrown error reflects this change, but the actual thrown error does not.

I wonder if the code runs from `__pycache__` (compiled bytecode), while symbols are shown from actual scripts (`.py`) files, which are out of sync.

> your test functions DO NOT raise the error!

2) If you meant my initial function, it shouldn't raise any errors. If you meant the mock functions - they might, and it's useful to know that they don't.

3) I see your point about distinct classes not raising an error. The issue is, regardless of what classes are used, I can't see where this error would be raised, under ANY circumstances.

Would you mind sharing some `json` or entire code snippet that produces the error? Alternatively, what also helps is learning the values of `target_classes` inside the library, right at the point the error happens.

**Reply by LinasKo** [reply]
To confirm: on my end, I am able to merge datasets with both overlapping and non-overlapping classes, with no errors, using either mocked json files and your code, or Universe datasets an my mock.

### Json
Merging these 3:

<details>
<summary>Json 1</summary>

```json
[
    {
      "image": "data:image/jpg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/",
      "bbox": [
        {
          "xmin": 0.1,
          "ymin": 0.1,
          "xmax": 0.3,
          "ymax": 0.4,
          "class": "PalletA",
          "id": "image_0001"
        }
      ],
      "height": 600,
      "width": 800
    }
  ]
```

</details>

<details>
<summary>Json 2</summary>

```json
[
    {
      "image": "data:image/jpg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/",
      "bbox": [
        {
          "xmin": 0.2,
          "ymin": 0.2,
          "xmax": 0.4,
          "ymax": 0.5,
          "class": "PalletA",
          "id": "image_0002"
        }
      ],
      "height": 720,
      "width": 1280
    }
  ]
```
</details>

<details>
<summary>Json 3</summary>

```json
[
    {
      "image": "data:image/jpg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/",
      "bbox": [
        {
          "xmin": 0.05,
          "ymin": 0.05,
          "xmax": 0.2,
          "ymax": 0.25,
          "class": "PalletB",
          "id": "image_0003"
        }
      ],
      "height": 480,
      "width": 640
    }
  ]
  
```
</details>

<details>
<summary>code</summary>

```python
import os
import json
import base64
import numpy as np
import supervision as sv
from collections import defaultdict

IMAGE_READ_INDEX = 0
DATA_DIR_PATH = "./json"


def json_file_to_detection_dataset(_json_file_path: str, _image_read_index: int) -> sv.DetectionDataset:

  # Check if the JSON file exists.
  if not os.path.exists(_json_file_path):
      print(f"Error: {_json_file_path} not found")
  try:
      # Open and load the JSON file
      with open(_json_file_path, 'r') as f:
          json_data = json.load(f)
  except json.JSONDecodeError as e:
      # Print an error message if there is a problem decoding the JSON.
      print(f"Error decoding JSON from {_json_file_path}: {e}")

  annotations_as_detections = []
  image_paths_list = []
  classes = []

  for image_item in json_data:
    # Build the current custom annotations dictionary by which more labelling information.
    # from a single image item within the json file can be passed onto the DetectionDataset.
    # Apparently a defaultdict(list) is required. A regular dictionary does not suffice!
    current_custom_annotations = defaultdict(list)

    # Decode the base64 image data.
    image_data = base64.b64decode(image_item['image'].split(',')[1])
    image_path = os.path.join(DATA_DIR_PATH, f"image_{_image_read_index}.jpg")
    image_paths_list.append(image_path)
    with open(image_path, "wb") as f:
      f.write(image_data)

    # Extract bounding box information.
    xmin = image_item["bbox"][0]["xmin"]
    ymin = image_item["bbox"][0]["ymin"]
    xmax = image_item["bbox"][0]["xmax"]
    ymax = image_item["bbox"][0]["ymax"]
    current_class = image_item["bbox"][0]["class"].split("#")[-1]

     # We assume all the JSON files only have "PalletA" class
    current_class_id = 0 #if current_class == "PalletA" else -1 #Assign -1 if class is not PalletA to filter out later if needed.

    current_image_id = image_item["bbox"][0]["id"].split("/")[-1]
    current_height = image_item["height"]
    current_width = image_item["width"]
    # ...
    # There is even more bounding box information to be found in the jsons.

    current_custom_annotations["class_name"].append(current_class)
    current_custom_annotations["image_id"].append(current_image_id)
    current_custom_annotations["height_in_pixels"].append(current_height)
    current_custom_annotations["width_in_pixels"].append(current_width)

    # Depending on the size of the input image, the xyxy values should be scaled.
    current_xyxy = [xmin * int(current_width), ymin * int(current_height), xmax * int(current_width), ymax * int(current_height)]
    classes.append(current_class)

    # For every image, annotations are combined to a sv.Detections object
    annotations_as_detections.append(
        sv.Detections(
            xyxy=np.array([current_xyxy], dtype=np.float32),
            class_id=np.array([current_class_id], dtype=int),
            data=dict(current_custom_annotations)
        )
    )
    # Increase the global image index for reading json images.
    _image_read_index += 1

  # Keep only the unique classes!
  classes = list(set(classes))

  output_detection_dataset = sv.DetectionDataset(classes, image_paths_list, dict(zip(image_paths_list, annotations_as_detections)))
  return output_detection_dataset, _image_read_index

def merged_detection_dataset_from_directory(_directory_path, _image_read_index):
  # Now we generate a DetectionDataset for every json file and merge them.
  json_filenames = []
  for filename in os.listdir(_directory_path):
      if filename.endswith(".json"):  # Check for JSON extension
          json_filenames.append(filename)

  print(f"Found JSON files: {json_filenames}")

  if json_filenames:
    datasets_to_merge = []
    all_classes = []
    for json_file_name in json_filenames:
        current_detection_dataset, _image_read_index = json_file_to_detection_dataset(os.path.join(_directory_path, json_file_name), _image_read_index)
        datasets_to_merge.append(current_detection_dataset)

    # Merge the datasets.
    merged_detection_dataset = sv.DetectionDataset.merge(datasets_to_merge)

  return merged_detection_dataset, _image_read_index

# Create the training and the testing detection datasets.
detection_dataset_train, IMAGE_READ_INDEX = merged_detection_dataset_from_directory("./json", IMAGE_READ_INDEX)
detection_dataset_test, IMAGE_READ_INDEX = merged_detection_dataset_from_directory("./json", IMAGE_READ_INDEX)

print(detection_dataset_train)
print(detection_dataset_test)
```

</details>

### Universe datasets

<details>
<summary>Code</summary>

```python
import os
from roboflow import Roboflow
import supervision as sv

key = os.getenv("ROBOFLOW_API_KEY")
assert key is not None, "ROBOFLOW_API_KEY is not set"
rf = Roboflow(api_key=key)


def find_class_mapping(dataset: sv.DetectionDataset):
    class_names = set()
    class_index_name_pairs = set()


    # Existing class names
    for detections in dataset.annotations.values():
        for class_index, class_name in zip(detections.class_id, detections.data.get("class_name", [])):
            class_index_name_pairs.add((class_index, class_name))
    
    # All class names
    for detections in dataset.annotations.values():
        for class_name in detections.data.get("class_name", []):
            class_names.add(class_name)
    return class_index_name_pairs, class_names


def load_dataset(ws_name, project_name, version_number):
    project = rf.workspace(ws_name).project(project_name)
    version = project.version(version_number)
    rf_dataset = version.download("yolov8")
    dataset = sv.DetectionDataset.from_yolo(
        images_directory_path=f"{rf_dataset.location}/train/images",
        annotations_directory_path=f"{rf_dataset.location}/train/labels",
        data_yaml_path=f"{rf_dataset.location}/data.yaml",
    )
    return dataset

ds_1 = load_dataset("abcmvs", "ops-pallet-detection", 1)
ds_2 = load_dataset("karthik-arumugham", "ops-pallet-detection-jfbop", 1)
ds_3 = load_dataset("ahn-axmvg", "iso-okprx", 1)

print(find_class_mapping(ds_1))
print(find_class_mapping(ds_2))
print(find_class_mapping(ds_3))
                
merged = sv.DetectionDataset.merge([ds_1, ds_2, ds_3])

print(merged)
```

</details>

**Reply by bmorlo** [reply]
I have the following update for you. It looks like that I did not load your updated supervision library correctly.

Now, when using your ``git+https://github.com/roboflow/supervision.git@fix/dataset-merge-error`` package I do not run into errors at all. If, however, I use the original ``supervision`` package, the error persists.

Not sure why this is. As far as I can tell, you did not change much in the updated directory, right?

**Reply by LinasKo** [reply]
Good news.

In the last update we made merging treat classes as case-sensitive. This may have inadvertently fixed the bug you're witnessing 😄

https://github.com/roboflow/supervision/pull/1643/files  

Which supervision do you have when the error is thrown? You can find out with `pip freeze | grep supervision`

**Reply by bmorlo** [reply]
Yeah, I really believe I was simply using an outdated version (0.22). Now when I install and work with the newest version, everything runs smoothly.

I hope I understand you correctly. I had only one class called "PalletA". No e.g. "palleta" class or something. Hence, I'm not so sure about this being an issue of case sensitivity.

**Reply by LinasKo** [reply]
Oh, the lower-casing was applied before the datasets were merged.

Step 1: lower-cased classes from all datasets were merged into a list. ['palleta', 'palletb', 'palletc'].
Step 2: it would immediately try to look for 'PalletA' in this list, and throw the error you're seeing.

This is exactly why the bug would happen. Glad to have it sorted at last.

**Reply by bmorlo** [reply]
Ah, now that makes sense! Thank you so much for your amazing support! All the best.

**Reply by LinasKo** [reply]
Glad to help!
